### PR TITLE
Menu for Angular components lib

### LINKED ISSUE #140: Angular Menu
This is Angular counterpart to https://github.com/AgnosticUI/agnosticui/issues/88

### PR DESCRIPTION
**Pull request by vitale232**
# Pull Request 

This can still be considered a WIP

## Description

Implements `AgnosticUI` `Menu` for Angular component library

Fixes #140 

## Checklist:

Please delete options that are not relevant.

- [x] Is this a new feature
- [x] Does this add a new component? If so, have you updated all framework's `copystyles.js` to pull in said CSS and verified it's working properly?
- [x] Storybook passing?
- [ ] Tests passing?
- [x] Have you added to the framework's /examples apps? e.g. `agnosticui-react/examples`? These are kitchen sink sanity checks which help to verify a component is working and derive documentation snippets.
- [ ] Have you updated the docs? These live in [site/docs](https://github.com/AgnosticUI/agnosticui/tree/master/site/docs)

### CODE REVIEW

**Comment by vitale232** [comment]
The biggest open issue in my mind is the API for the `menuItems`. Currently, we're just accepting an array of `MenuItem`s, and rendering the buttons under-the-hood on behalf of app devs.

```html
    <ag-menu class="my-menu"
             menuTitle="Default Menu"
             [menuItems]="[{label: 'Item 1' }, {label: 'Item 2'}, {label: 'Item 3'}]"></ag-menu>
```

Other popular Angular libraries like [Material](https://material.angular.io/components/menu/examples#menu-icons) allow devs to pass content into the menu component, which we do not currently support. 

Which API is more similar to the other `AgnosticUI` implementations? The biggest hurdle on the implementation side to have an API more similar to material is the keyboard accessibility requirements defined by `ag-ui`.

Any thoughts here?

**Review by vitale232 — COMMENTED** [review]
The :host selector in MenuCompnents decorator should have been removed with the cleanup commit

**Comment by roblevintennis** [comment]
I'm out with family for most of the day but will try to find to to look. From what you mentioned I agree it's ideal if consumers can project children content.

I forget for certain but I thought AgnosticUI Angular Tabs component might have tackled that. Have you tried looking at that component?

**Comment by vitale232** [comment]
@roblevintennis No worries on response times from my perspective, I'm balancing personal life and day job as well and totally get it 😅 

So after some more consideration, I'm pretty sure this is the API I'd like to achieve. What do you think of this?

```html
<ag-menu class="my-menu"
         menuTitle="Default Menu">
  <button ag-menu-item
          *ngFor="let item of ['Item 1', 'Item 2', 'Item 3']; let idx = index"
          [disabled]=" false"
          [isSelected]="idx === selectedMenuItemIndex"
          (click)="selectedMenuItemIndex = idx;">
    {{ item }}
  </button>
</ag-menu>
```

As you can see, I've moved the responsibility of setting the `menu-item-selected` class on the menu item out of the library and on to the consumer. This is pretty different than some of the other implementations of AG Menu from what I can tell, but I actually think it's a bit more flexible. Lots (maybe most?) uses for a menu don't necessarily need to persist the previous menu selection, but lib users would be able to achieve that when it's a requirement.

What do you think of this API change? Does something like the above look reasonable? One concern looking at the above is that some users may default to a `<div ag-menu-item>` rather than a button, but we could probably enforce that menu items' must be buttons in the selector and documentation. If we like this API and thought process, I will likely abandon this PR and try another implementation... 

The tabs will be a good starting point. The biggest additional complexity over the tabs is that we'll need to be aware of the events that occur on the menu items, whereas I believe the Tabs do not need to be aware of any events that occur on the TabsPanel. Does that make sense, or am I thinking about this wrong?

**Comment by roblevintennis** [comment]
Ok, long comment coming :-)

So, my question for all this was again if we need to be able to project custom content and I had forgot whether or not we're supporting in the other frameworks. I think we are so I think an API like:

```js
*ngFor="let item of ['Item 1', 'Item 2', 'Item 3']; let idx = index"
```

is limiting the menu option items to be strings that the library is assembling elements around. So if they want to add an icon beside the string or similar they're not able to. I thought this was something you had mentioned Material allowed for too no? So I think the API needs to provide for this in Angular too.

I think the solution commonly in Angular to custom projection in https://angular.io/api/common/NgTemplateOutlet -- again, this is basically what was used for the Angular Tabs…

___

## Oops -- Tabs with Custom Projected Content Ain't Working!

So I went to confirm all this and realized Angular Tabs was borked 🤦 

Here's the bug trying to grab ng-content with view children but that's for content children as I understand from reading some docs just now:

image


 So I fixed it just now in

```shell
[master 710e5c90] Fix the bug in Angular Tabs
```

You can see from the console.log output I'm using my arrow keys (once I've "tabbed into" the tab group) to navigate in a circular fashion and now it's working properly:

image

___

Ok, back to the Menu API requirements. I think we DO want to be able to project custom content (so back to replicated the Tab above).

## You really don't have to read anything below, it's just a bunch of proof that the other frameworks support custom projected content via react render props, slots, etc.

___

TL;DR other framework Menu API's support custom content

Here's randomly selected snippets from AgnosticUI's various frameworks that I found just now to confirm they're all providing for custom content projection:

Vue

```vue
          <Menu
            id="menu2"
            menuTitle="Players (!closeOnClickOutside)"
            :closeOnClickOutside="false"
          >
            <template #menuitem-1>Menu Item 1</template>
            <template #menuitem-2>Menu Item 2</template>
            <template #menuitem-3>Menu Item 3</template>
            <template #menuitem-4>Menu Item 4</template>
            <template #menuitem-5>Menu Item 5</template>
            <template #icon>&#9662;</template>
          </Menu>
```

Svelte

```js
  /**
   * Menu
   */
  const menuItems = [
    {
      label: "Menu Item 1",
      menuItemComponent: MenuItem
    },
    {
      isDisabled: true,
      label: "Menu Item 2",
      menuItemComponent: MenuItem
    },
    {
      label: "Menu Item 3",
      menuItemComponent: MenuItem
    },
    {
      label: "Menu Item 4",
      menuItemComponent: MenuItem
    },
    {
      label: "Menu Item 5",
      menuItemComponent: MenuItem
    },
  ]
...
      <Menu
        menuItems={menuItems}
        id="mymenu1"
        menuTitle="Players"
        onOpen={(selectedItem) => console.log('onOpen - selectedItem: ', selectedItem)}
        onClose={() => console.log('onClose called...')}
      />
```

and in menu component (svelte)

```html
  <div class="menu-items" id={id} role="menu" hidden={!expanded}>
    {#each menuItems as item, i}
      <svelte:component
        this={item.menuItemComponent}
        bind:this={menuItemRefs[i]}
        classes={menuItemClasses(selectedItem === i)}
        isSelected={selectedItem === i}
        disabled={item.isDisabled}
        on:click={onMenuItemClicked(i)}
        on:keydown={(ev) => onMenuItemKeyDown(ev, i)}
      >
      {item.label}
    </svelte:component>
    {/each}
  </div> 
```

And React interface takes react elements (so you're essentially able to arbitrarily project the items):

```
menuItems: JSX.Element[];
```

and then in the react template jsx:

```jsx
<div className={styles.items} id={id} role="menu" hidden={!expanded}>
        {menuItems.map((btn, i) => cloneElement(
          btn,
          {
            key: i,
            index: i,
            isSelected: selectedItem === i,
            isDisabled: btn.props.isDisabled,
            ref: itemRefs.current[i],
            className: itemClasses,
            onClick: () => {
              selectItem(i);
              if (closeOnSelect) {
                setOpened(false);
                focusTriggerButton();
              }
            },
            onKeyDown: (ev: KeyboardEvent<HTMLElement>) => onMenuItemKeyDown(ev, i),
            children: btn.props.children,
          },
          btn.props.children,
        ))}
      </div>
```

**Comment by roblevintennis** [comment]
Ok, so next steps? I see two choices

1. We start with what you have and document Angular Menu doesn't yet support custom content project (but will soon) and make another ticket
2. You just keep iterating here and get that working

I know it's frustrating to keep iterating and you've done a lot of work so I don't want you're morale to suffer :) The only thing if we do option 1 is I still do need to find time to pull down your PR and verify (just as a matter of due dilligence). But if you're leaning towards number 2 then I won't waste time doing that.

Lmk if all my feedback above is understandable and fair to you! I definitely want to support you through to the finish line and appreciate your time and contribution here!

**Comment by vitale232** [comment]
Hey, thanks for taking the time to write this up. I appreciate the support!

The review of the other APIs helps a lot. I have interest in other frameworks, but have only worked professionally in Angular. It's cool to see what you're up to there as well, and it helps ground things!

I def think option 2 works for me. My fav part of library code is that you get to keep iterating :D In fact, I spent some time working on a different approach today. I think I'm closer to an impl that allows bring-your-own-content. I would _love_ to abstract away the need for an `ng-template`, but I'm not seeing a way at the moment.

What do you think of something like this? It's a PR in my fork of the lib. The link should point you to the kitchen sink where the Menu is being used three different ways. I personally think this is a lot closer

https://github.com/vitale232/agnosticui/pull/1/files#diff-7756f63413f82e67be6515b819c20fbec312afc0b6ce8a452ca212cccdcfcddc

If we're thinking this is something worth cleaning up, the next step will be to get Storybook upa nd running again.

**Comment by roblevintennis** [comment]
I just looked at the diff from my phone and I like the direction where you have the list of labels on the template ref button (menu item) and loop those with the `{{ item }}` 😀😀 

Proving consumer can project custom content 👍🏽🔥

Angular is honestly the framework I'm least skilled in and I just don't love the verbosity and I'm horrible at RxJs so I'm hoping AgnosticUI becomes popular enough that we grow a small team of core contributors and someone much better then I can own the Angular package 🤓

I do Angular at my day job and I still suck at it almost a year in 🤣🤣🤣

**Comment by vitale232** [comment]
Hey, whatever pays the bills! Having an affinity towards rxjs makes it difficult for me to leave Angular. Totally hear you though, it's like a mini dotnet. One of the things that made it click for me was hearing the origin story on a podcast. Long story short, the original idea was to _extend_ HTML for designers. It's a come long way from that, but it made concepts like directives click for me and made me focus more on the native platform when I can.

Speaking of paying the bills, I've been super swamped at the day job. I'm going to close this out for now. I'll clean up my fork, rehydrate it with your latest work, and make a new PR. Hopefully I'll get some time this weekend to work on it, and we can land this menu!

**Comment by roblevintennis** [comment]
Thanks yes it's not RxJs it's me 🤣 hard to teach old dogs new tricks
